build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/core/jsoncdb/generator/CompilationDatabaseContributionManager.java

```diff
@@ -0,0 +1,137 @@
+/********************************************************************************
+ * Copyright (c) 2023, 2024 Renesas Electronics Corp. and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ********************************************************************************/
+package org.eclipse.cdt.managedbuilder.core.jsoncdb.generator;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.cdt.managedbuilder.core.IConfiguration;
+import org.eclipse.cdt.managedbuilder.core.IToolChain;
+import org.eclipse.cdt.managedbuilder.core.jsoncdb.CompilationDatabaseInformation;
+import org.eclipse.cdt.managedbuilder.core.jsoncdb.ICompilationDatabaseContributor;
+import org.eclipse.core.runtime.Assert;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IConfigurationElement;
+import org.eclipse.core.runtime.IExtension;
+import org.eclipse.core.runtime.IExtensionPoint;
+import org.eclipse.core.runtime.Platform;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.jdt.annotation.NonNull;
+
+final class CompilationDatabaseContributionManager {
+	private static final String ATTRIB_RUNNER = "runner"; //$NON-NLS-1$
+	private static final String ATTRIB_TOOLCHAIN_ID = "toolchainID"; //$NON-NLS-1$
+	private static final String ID_COMPILATIONDATABASE = "compilationDatabase"; //$NON-NLS-1$
+	private static final String EXTENSION_ID = "compilationDatabaseContributor"; //$NON-NLS-1$
+	@NonNull
+	private final Map<String, ICompilationDatabaseContributor> loadedInstances;
+
+	private class EmptyCompilationDatabaseContributor implements ICompilationDatabaseContributor {
+
+		@Override
+		public final @NonNull List<CompilationDatabaseInformation> getAdditionalFiles(@NonNull IConfiguration config) {
+			return Collections.emptyList();
+		}
+	}
+
+	private Map<String, IConfigurationElement> factoryExtensions;
+	private static CompilationDatabaseContributionManager instance;
+
+	private CompilationDatabaseContributionManager() {
+		factoryExtensions = new HashMap<>();
+		loadedInstances = new HashMap<>();
```

Comment:
I marked previous comments resolved around these fields because they weren't appearing on the latest patchset.

These fields lifecycle is unclear, made more complicated by the local variable with a similar name in initialise method.

Can these fields be final, assigned at declaration and have a clearer lifecycle? If my comment is unclear, let me know and I will provide a commit to explain it.

Reply:
This is part of what I meant - I will add a commit to this PR to explain explicitly what I mean.

Reply:
I added b6c8e30b7e7cd2dbeeeb00bd8d7ed56c86f26c3c which shows what I meant. It is fairly minor, but removes some code and gets rid of the extra temporary map.